### PR TITLE
Convex Strategy

### PR DESCRIPTION
**Pull request by DanielVF**
From issue #623, add Convex/Curve support

- [x] Merge in all audit changes to old 3Pool strategy
- [x] Upgrade language version
- [x] Simulate resistance to Curve price attacks

Contract change checklist:
  - [x] Code reviewed by 2 reviewers. See [template](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review.md) and [example](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review-Example.md).
  - [x] Unit tests pass
  - [x] Slither tests pass with no warning
  - [x] Echidna tests pass if PR includes changes to OUSD contract (not automated, run manually on local)

### CODE REVIEW

File: contracts/contracts/mocks/MintableERC20.sol [hunk 1]
```diff
@@ -20,4 +22,15 @@ contract MintableERC20 is IMintableERC20, ERC20 {
...
         return true;
     }
+
+    /**
+     * @dev Function to burn the minted tokens
+     * @param account The address with the tokens to burn.
+     * @param value The amount of tokens to burn.
+     * @return A boolean that indicates if the operation was successful.
+     */
+    function burn(address account, uint256 value) public returns (bool) {
```
Comment (tomlinton) [inline contracts/contracts/mocks/MintableERC20.sol:32]: We have MintableERC20.sol and BurnableERC20.sol, although the burn function has a different signature. We should try clean this up somehow, either using the existing burn function or add another with this signature.

**Comment by DanielVF** [comment]
I've changed the checkBalance function to work off virtual_price(). This gets us the "up only" behavior that we want from our strategies. 

However, with this simplified divide by three method it does mean that the proportions don't match the proportions in 3pool.In production this should not matter. The numbers shown are the actual numbers we could receive, but we could get more coins by doing a withdrawals that matches the proportions.

```
if (totalPTokens > 0) {
            uint256 virtual_price = curvePool.get_virtual_price();
            uint256 value = (totalPTokens * virtual_price) / 1e18;
            uint256 assetDecimals = Helpers.getDecimals(_asset);
            balance = value.scaleBy(assetDecimals, 18) / 3;
        }
```

**Comment by DanielVF** [comment]
## Convex testing

```python
from world import *

# same as
# brownie.network.web3.provider.make_request('hardhat_impersonateAccount', [STRATEGIST])
unlock(STRATEGIST)

## Before:
print(vault_core.totalValue() / 1e18)

## 1. Funds move in
vault_admin.reallocate(COMP_STRAT, CONVEX_STRAT, [usdc], [1e12], {'from': STRATEGIST})
print(vault_core.totalValue() / 1e18)

## 2. Funds move out
vault_admin.reallocate(CONVEX_STRAT, COMP_STRAT, [usdc], [int(0.9*1e12)], {'from': STRATEGIST})

## 3. Funds move back in
vault_admin.reallocate(COMP_STRAT, CONVEX_STRAT, [usdc], [1e12], {'from': STRATEGIST})
print(vault_core.totalValue() / 1e18)

## 4. Harvest
vault_admin.harvest({'from': STRATEGIST})
show_transfers(history[-1])

## 5. Withdraw all
vault_admin.withdrawAllFromStrategy(CONVEX_STRAT, {'from': STRATEGIST})
print(vault_core.totalValue() / 1e18)
```


## Test with 3Pool price bending

Our holdings should go up during this test, since the attacker will be sending a lot of fees to pool.

The important bit of this test is that regardless of what the attacker does, our total value **only goes up**.

```python
from world import *
unlock(STRATEGIST)

## Almost all funds to 3 pool
vault_admin.withdrawAllFromStrategy(CONVEX_STRAT, {'from': STRATEGIST})
vault_admin.reallocate(COMP_STRAT, CONVEX_STRAT, [usdc, usdt, dai], [3*1e12, 3*1e12, 9*1e18*1e6], {'from': STRATEGIST})

## Setup Attacker
unlock(USDT_BAGS)
usdt.approve(threepool, 1e70, {'from': USDT_BAGS})
dai.approve(threepool, 1e70, {'from': USDT_BAGS})

## Before
print(vault_core.totalValue() / 1e18)

## Billion dollars buy DAI
threepool.exchange(2, 0, 1000 * 1e6 * 1e6, 0, {'from': USDT_BAGS})
print(vault_core.totalValue() / 1e18)

## Unbend
threepool.exchange(0, 2, 600 * 1e6 * 1e18, 0, {'from': USDT_BAGS})
print(vault_core.totalValue() / 1e18)

## Billion dollars buy DAI
threepool.exchange(2, 0, 1000 * 1e6 * 1e6, 0, {'from': USDT_BAGS})
print(vault_core.totalValue() / 1e18)

## Unbend
threepool.exchange(0, 2, 500 * 1e6 * 1e18, 0, {'from': USDT_BAGS})
print(vault_core.totalValue() / 1e18)
```

File: contracts/deploy/022_convex.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        .connect(sDeployer)
+        .transferGovernance(governorAddr, await getTxOpts())
+    );
+
+    // Governance Actions
+    // ----------------
+    return {
+      name: "Switch to new Convex strategy",
+      actions: [
+        // 1. Accept governance of new AAVEStrategy
```
Comment (franckc) [inline contracts/deploy/022_convex.js:88]: typo: 3pool, not aave

**Review by franckc — APPROVED** [review]
This makes the 3pool strategy code much simpler! nice.

I did not do any local testing. But overall the code looks solid.

One comment would be to update the hardhat debug task that we use to dump the state of the system to include convex specific data: https://github.com/OriginProtocol/origin-dollar/blob/master/contracts/tasks/debug.js
This can be done as a separate PR.

LGTM

**Comment by franckc** [comment]
Adding @crazybuster to this PR - the more eyes the better on any new strategy!

**Comment by DanielVF** [comment]
Our withdrawAll() minimum amount check has been somewhat meaningless, since the strategy calculates it on-chain using same on-chain data that 3Pool will be using. This was also pointed out during the audit.

With the change to using virtual price to get the strategy's balances of each stablecoin, the previous method we we were using for minimum amount check no longer worked, since it would now be checking for an almost perfect 1/1/1 ratio of coins in 3pool, which almost never happens.

Given that is doesn't provide any benefit here, and just duplicating work that the 3Pool contract is already doing, [I've removed it](https://github.com/OriginProtocol/origin-dollar/pull/768/commits/0a759fdcadbd36126fa6dabc63f4783508b157ac).

In the long run, the correct way to check the financial results of reallocaton actions, is at the vault level, using data computed off-chain and sent along with the reallocaton transaction.

**Comment by franckc** [comment]
@tomlinton Do you mind doing a final review? Then we'll deploy this puppy:)

File: contracts/deploy/029_convex.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
+const { deploymentWithProposal } = require("../utils/deploy");
+
+module.exports = deploymentWithProposal(
+  { deployName: "022_convex" },
```
Comment (tomlinton) [inline contracts/deploy/029_convex.js:4]: Rename to 029_

**Review by tomlinton — APPROVED** [review]

**Comment by Qusai777** [comment]
This is big!

**Pull request merged**